Subprocess compile

aorenste opened the pull request:
Add a mode to `fx_codegen_and_compile()` to compile in a separate process. This is to prepare for async compile where we'll compile and run eager in parallel (and also be able to move the compile phase to a remote computer).

Added a test based which runs the test_torchinductor tests with subprocess compiling turned on.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #146134



cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @desertfire @chauhang @aakhundov

aorenste committed:
WIP: subprocess compile

[ghstack-poisoned]

aorenste committed:
Update on "WIP: subprocess compile"




[ghstack-poisoned]

aorenste committed:
Update on "WIP: subprocess compile"




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste committed:
Update on "WIP: subprocess compile"




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste committed:
Update on "WIP: subprocess compile"




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste committed:
Update on "WIP: subprocess compile"




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

jamesjwu commented on torch/_inductor/virtualized.py:
```diff
@@ -95,6 +95,11 @@ class NullHandler:
 
+# If a virtualized value is set to _PoisonedVirtual then any attempt to get the
+# value will result an an exception being raised.
+_PoisonedVirtual = object()
```
Let's add a bit more to the comment to explain when poisoning a virtualized is useful

jamesjwu commented on torch/_inductor/compile_fx.py:
```diff
@@ -1383,6 +1632,172 @@ def _send_to_child(
+    def _postprocess(self, output: _WireProtocolOutput) -> None:
+        # Since our metrics were gathered in a subprocess make sure to add them
+        # here.
+        CachedMetricsHelper.apply_deltas(output.metrics)
```
Which specific metrics do we expect to apply here? Does it include counters? CompilationMetrics?

aorenste replied:
I added the metrics that seemed to be checked by tests. We certainly could/should add any metrics needed in the future.

jamesjwu replied:
There's tests in test_structured_trace for structured logs, fwiw, once we get there

jamesjwu commented on torch/_inductor/compile_fx.py:
```diff
@@ -1383,6 +1632,172 @@ def _send_to_child(
+
+        f = pool.submit(_SubprocessFxCompile._run_in_child_subprocess, input, extra_env)
+        last = time.time()
+        while not f.done():
```
Is a poll based approach really the only way to wait for a child to finish? I would have expected something nicer like a Future.wait() or something...

jamesjwu commented on torch/_inductor/compile_fx.py:
```diff
@@ -1383,6 +1632,172 @@ def _send_to_child(
+        # incremented when we clear the cache, send the clock to the worker and
+        # only clear caches if the clock changed since last time.
+        #
+        clear_inductor_caches()
```
Can you explain why we need to clear the inductor caches in the subprocess? Why can't the child simply use the inductor cached results on disk? And if the child is on RE/elsewhere, why would the on disk cache be affected?

aorenste replied:
There's a long comment immediately preceding this line which tries to explain the rationale.  You're right that in a hermetic RE worker it shouldn't matter - but for a subprocess we need some way to coordinate that the parent called clear_inductor_caches() with the child.

This is another thing we need to respect for tests to pass...

jamesjwu commented on torch/_inductor/compile_fx.py:
```diff
@@ -1383,6 +1632,172 @@ def _send_to_child(
+        clear_inductor_caches()
+        torch._inductor.metrics.reset()
+
+        # TODO: turn off config.fx_graph_async_compile
```
Yeah speaking of which, not sure what the behavior here is if TORCHINDUCTOR_COMPILE_THREADS > 1, since the child would be spawning more processes. We can maybe just assume that AsyncCompile is off for now

jamesjwu commented on torch/_inductor/compile_fx.py:
```diff
@@ -1283,6 +1378,119 @@ def deserialize(self, constants: CompiledFxGraphConstants) -> _WireProtocolOutpu
+        self.captured_logs.remove()
+
+
+class _CapturedLogs:
```
Going to assume this does not cover structured trace logs. It may be good to think about how to pass that along the wire. 

If it's just a subprocess on the same machine, we could probably just initialize LazyTraceHandler to the same `dedicated_log_torch_trace_*` as previously held, and it would append correctly to that JSON whenever someone called `trace_structured`. But it's more likely you want to handle it similarly to how you handle this, where we want to capture calls to structured_trace similarly to stderr logs.

jamesjwu reviewed (commented):
Did a first pass, overall makes sense. I wonder if piggybacking off of some of the CodeCacheFuture stuff in codecache.py for handling waiting for the subprocess may be a good abstraction to tack on. Then we could use the same Futures to handle both parallel triton compile and this subprocess compile.

aorenste committed:
Update on "Subprocess compile"

Add a mode to `fx_codegen_and_compile()` to compile in a separate process. This is to prepare for async compile where we'll compile and run eager in parallel (and also be able to move the compile phase to a remote computer).

Added a test based which runs the test_torchinductor tests with subprocess compiling turned on.




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste committed:
Update on "Subprocess compile"

Add a mode to `fx_codegen_and_compile()` to compile in a separate process. This is to prepare for async compile where we'll compile and run eager in parallel (and also be able to move the compile phase to a remote computer).

Added a test based which runs the test_torchinductor tests with subprocess compiling turned on.




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste committed:
Update on "Subprocess compile"

Add a mode to `fx_codegen_and_compile()` to compile in a separate process. This is to prepare for async compile where we'll compile and run eager in parallel (and also be able to move the compile phase to a remote computer).

Added a test based which runs the test_torchinductor tests with subprocess compiling turned on.




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste committed:
Update on "Subprocess compile"

Add a mode to `fx_codegen_and_compile()` to compile in a separate process. This is to prepare for async compile where we'll compile and run eager in parallel (and also be able to move the compile phase to a remote computer).

Added a test based which runs the test_torchinductor tests with subprocess compiling turned on.




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste commented:
@pytorchbot merge

aorenste committed:
Update on "Subprocess compile"

Add a mode to `fx_codegen_and_compile()` to compile in a separate process. This is to prepare for async compile where we'll compile and run eager in parallel (and also be able to move the compile phase to a remote computer).

Added a test based which runs the test_torchinductor tests with subprocess compiling turned on.




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste committed:
Update on "Subprocess compile"

Add a mode to `fx_codegen_and_compile()` to compile in a separate process. This is to prepare for async compile where we'll compile and run eager in parallel (and also be able to move the compile phase to a remote computer).

Added a test based which runs the test_torchinductor tests with subprocess compiling turned on.




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

malfet commented:
@pytorchbot revert -m "looks like it broke slow jobs, see https://hud.pytorch.org/hud/pytorch/pytorch/e1dee4ccb3df4d5a196bff451c07863f2f4fc2d6/3?per_page=50&name_filter=slow%20%2F%20linux-focal-rocm" -c nosignal

aorenste commented:
Replaced by #148635
